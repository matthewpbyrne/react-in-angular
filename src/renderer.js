import React from 'react';
import ReactDOM from 'react-dom';

export function createRenderer (Component, container) {
    function render (props) {
        ReactDOM.render(<Component {...props} />, container);
    };

    function unmount () {
        ReactDOM.unmountComponentAtNode(container)
    };

    return {
        render,
        unmount
    };
}
