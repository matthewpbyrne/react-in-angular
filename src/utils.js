import React from 'react';

export function createMapScopeToProps (bindings) {
    return function mapScopeToProps (scope) {
        return Object.keys(bindings).reduce((props, propName) => {
            // We expect that properties which are bound with & are functions
            if (bindings[propName] === '&') {
                props[propName] = (event) => scope[propName]({ $event: event });
            } else {
                props[propName] = scope[propName];
            }

            return props;
        }, {});
    };
}

export function wrapWithDecorators (decorators) {
    const compose = (render, decorate) => {
        // We create a new render function which takes props from outside
        const renderDecorator = (props) => {
            // Function which takes the passed props to render the component
            const renderComponentWithProps = () => render(props);

            // Pass the render function for component rendering to the decorate function
            // as render function
            return decorate(renderComponentWithProps);
        };

        return renderDecorator;
    };

    return function withDecorators (Component) {
        const render = (props) => (<Component {...props} />);

        return decorators.reduce(compose, render);
    };
}