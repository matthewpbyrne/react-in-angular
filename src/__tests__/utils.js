import React from 'react';
import renderer from 'react-test-renderer';

import { createMapScopeToProps, wrapWithDecorators } from '../utils';

describe('createMapScopeToProps', () => {
    test('create mapScopeToProps functions ', () => {
        const bindings = {
            name: '<',
            onClickName: '&'
        };

        const mapScopeToProps = createMapScopeToProps(bindings);

        expect(typeof mapScopeToProps).toBe('function');
    });

    test('mapScopeToProps returns correct props', () => {
        const bindings = {
            name: '<',
            onClickName: '&'
        };
        const scope = {
            name: 'Hello',
            onClickName: jest.fn(),
            unknownProp: 'UNKNOWN'
        }
        const event = Symbol();
        const mapScopeToProps = createMapScopeToProps(bindings);

        const props = mapScopeToProps(scope);

        expect(props.name).toBe(scope.name);
        expect(typeof props.onClickName).toBe('function');

        props.onClickName(event)
        expect(scope.onClickName.mock.calls[0]).toEqual([ { $event: event } ]);
    })
});


describe('wrapWithDecorators', () => {
    test('returns function', () => {
        const decorators = [];

        expect(typeof wrapWithDecorators()).toBe('function');
    });

    describe('renders correctly', () => {
        const decorators = [
            (render) => (<span className="Decorator-1">{render()}</span>),
            (render) => (<span className="Decorator-2">{render()}</span>)
        ];
        const ComponentToDecorate = ({ content }) => (<span>{content}</span>);

        const Component = wrapWithDecorators(decorators)(ComponentToDecorate);

        const tree = renderer.create(
            <Component content="MyContent" />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});