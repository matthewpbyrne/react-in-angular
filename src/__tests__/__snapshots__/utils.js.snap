// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<span
  className="Decorator-2"
>
  <span
    className="Decorator-1"
  >
    <span>
      MyContent
    </span>
  </span>
</span>
`;
